cpp/ql/lib/semmle/code/cpp/commons/Printf.qll

```diff
@@ -990,7 +994,8 @@ class FormatLiteral extends Literal {
    * conversion specifier of this format string; has no result if this cannot
    * be determined.
    */
-  int getMaxConvertedLength(int n) {
+  int getMaxConvertedLength(int n, TFormattedSizeEstimateKind strategy, boolean strategyRelevant) {
+    (strategyRelevant = true or strategyRelevant = false) and // avoid "is not bound"...
```

Comment:
The "is not bound" warning is really there for a good reason: You're introducing a cartesian product: previously, the `getMaxConvertedLength` table would have 2 (ignoring the `this` argument) columns: `(result, n)`, with the interpretation that the `n`'th conversation can product a string of length at most `result`.

Now the table will look like `(result, n, strategy, strategyRelevant)`, where many results will have:
- A `(result, n, strategy, true)` row, and
- A `(result, n, strategy, false)` row

The `strategy` argument is actually also introducing a cartesian product, but unlike booleans, the compiler doesn't emit an error for this. If we're not in the `this.getConversionChar(n).toLowerCase() = "f"` case, then every previous row of the form `(result, n)` will now have both:
- A `(result, n, TFormattedSizeWithExtremeFloat(), strategyRelevant)` row, and
- A `(result, n, TFormattedSizeLimited(), strategyRelevant)` row.

We should either figure out how to make sure that we don't duplicate rows (i.e., by making sure that every argument is bound in every disjunction), or avoid the need for this change.

Reply:
I see, I will go back on this tomorrow morning, but this makes it more clear. I think I will try to make something similar and have no duplicate rows.

Reply:
Sounds good!